iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -434,6 +466,90 @@ public override async Task OpenAsync(CancellationToken cancellationToken)
                 _handlerSemaphore?.Release();
             }
         }
+        public override async void SetRefreshesOn(CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, cancellationToken, nameof(SetRefreshesOn));
+
+            DateTime refreshesOn = GetRefreshesOn();
+            if (refreshesOn < DateTime.MaxValue)
+            {
+                await StartLoopAsync(refreshesOn, cancellationToken);
+            }
+
+            if (Logging.IsEnabled)
+                Logging.Exit(this, cancellationToken, nameof(SetRefreshesOn));
+        }
+
+        private async Task StartLoopAsync(DateTime refreshesOn, CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, refreshesOn, nameof(StartLoopAsync));
+
+            if (_loopCancellationTokenSource == null
+                || _refreshLoop == null)
+            {
+                await StopLoopAsync();
+            }
+
+            _loopCancellationTokenSource = new CancellationTokenSource();
+            _refreshLoop = RefreshLoopAsync(refreshesOn, cancellationToken);
+
+            if (Logging.IsEnabled)
+                Logging.Exit(this, refreshesOn, nameof(StartLoopAsync));
+        }
+
+        private async Task RefreshLoopAsync(DateTime refreshesOn, CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, refreshesOn, nameof(RefreshLoopAsync));
+
+            TimeSpan waitTime = refreshesOn - DateTime.UtcNow;
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Info(this, refreshesOn, $"Before {nameof(RefreshLoopAsync)} with wait time {waitTime}.");
+
+                if (waitTime > TimeSpan.Zero)
+                {
+                    await Task.Delay(waitTime, cancellationToken).ConfigureAwait(false);
+                }
+
+                refreshesOn = await RefreshTokenAsync(cancellationToken).ConfigureAwait(false);
+
+                if (Logging.IsEnabled)
+                    Logging.Info(this, refreshesOn, $"After {nameof(RefreshLoopAsync)}");
```

Comment:
We're in RefreshLoopAsync, and we're not after it (still in the loop). Perhaps this log makes more sense?

```suggestion
                    Logging.Info(this, refreshesOn, "Token has been refreshed.");
```